File: cepi-ios/CEPITelemedicina/Chat/Episodios.swift
import Foundation

/// Qué página del hilo se mira: la consulta más nueva o una en particular. `String?` porque
/// los turnos viejos sin episodio también forman una página.
enum PaginaHilo: Equatable, Sendable {
    case masNueva
    case consulta(String?)
}

/// El hilo partido por consultas, como en IntakeChat.vue: un episodio por página.
struct Episodios: Sendable {
    /// Episodios distintos en orden de aparición. El activo va al final aunque todavía no tenga
    /// mensajes: una consulta recién abierta también es una página.
    let orden: [String?]

    init(mensajes: [MensajeHilo], activo: String?) {
        var vistos = Set<String?>()
        var orden: [String?] = []
        for mensaje in mensajes where vistos.insert(mensaje.episodio).inserted {
            orden.append(mensaje.episodio)
        }
        if let activo, !vistos.contains(activo) { orden.append(activo) }
        self.orden = orden
    }

    func indice(de pagina: PaginaHilo) -> Int {
        if case .consulta(let id) = pagina, let indice = orden.firstIndex(of: id) { return indice }
        return max(orden.count - 1, 0)
    }

    /// Los mensajes de una página. El aviso "Paciente activo: …" lo repite el bot en cada
    /// activación; se deja solo el último para que no se acumule.
    func visibles(_ mensajes: [MensajeHilo], indice: Int) -> [MensajeHilo] {
        var lista = orden.count <= 1 ? mensajes : mensajes.filter { $0.episodio == orden[indice] }
        if let ultimo = lista.lastIndex(where: Self.esAvisoDeActivacion) {
            lista = lista.enumerated()
                .filter { $0.offset == ultimo || !Self.esAvisoDeActivacion($0.element) }
                .map(\.element)
        }
        return lista
    }

    /// Solo en la consulta activa se escribe; las anteriores son de lectura.
    func esActiva(indice: Int, activo: String?) -> Bool {
        guard orden.count > 1 else { return true }
        let actual = activo ?? orden[orden.count - 1]
        return orden[indice] == actual
    }

    func etiqueta(indice: Int, mensajes: [MensajeHilo]) -> String {
        // Con una sola consulta, decirlo explica por qué las flechas no llevan a ningún lado.
        guard orden.count != 1 else { return "Única consulta" }
        guard !orden.isEmpty else { return "Sin consultas todavía" }
        let fecha = mensajes.first { $0.episodio == orden[indice] }?.fecha
        let cuando = fecha.map {
            " · " + $0.formatted(.dateTime.day(.twoDigits).month(.abbreviated).locale(Locale(identifier: "es")))
        } ?? ""
        return "Consulta \(indice + 1)/\(orden.count)\(cuando)"
    }

    /// El autor sobre un mensaje ajeno, una vez por racha del mismo autor (como un grupo de
    /// WhatsApp). `nil` = no se muestra.
    static func autor(en lista: [MensajeHilo], indice: Int) -> String? {
        let mensaje = lista[indice]
        guard !mensaje.propio else { return nil }
        if indice > 0 {
            let previo = lista[indice - 1]
            if !previo.propio && previo.autorId == mensaje.autorId { return nil }
        }
        return mensaje.esBot ? "Asistente" : (mensaje.autorNombre ?? "Profesional")
    }

    static func esAvisoDeActivacion(_ mensaje: MensajeHilo) -> Bool {
        mensaje.esBot && mensaje.contenido.drop(while: \.isWhitespace).hasPrefix("Paciente activo:")
    }
}

File: cepi-android/app/src/main/java/ec/cepi/telemedicina/chat/Episodios.kt
package ec.cepi.telemedicina.chat

import ec.cepi.telemedicina.api.Fechas
import ec.cepi.telemedicina.api.MensajeHilo

/** Qué página del hilo se mira: la consulta más nueva o una en particular. */
sealed interface PaginaHilo {
    data object MasNueva : PaginaHilo

    /** `null` porque los turnos viejos sin episodio también forman una página. */
    data class Consulta(val id: String?) : PaginaHilo
}

/** El hilo partido por consultas, como en IntakeChat.vue: un episodio por página. */
class Episodios(mensajes: List<MensajeHilo>, activo: String?) {
    /**
     * Episodios distintos en orden de aparición. El activo va al final aunque todavía no tenga
     * mensajes: una consulta recién abierta también es una página.
     */
    val orden: List<String?>

    init {
        val distintos = LinkedHashSet<String?>()
        mensajes.forEach { distintos += it.episodio }
        if (activo != null && activo !in distintos) distintos += activo
        orden = distintos.toList()
    }

    fun indice(pagina: PaginaHilo): Int {
        if (pagina is PaginaHilo.Consulta) {
            val indice = orden.indexOf(pagina.id)
            if (indice >= 0) return indice
        }
        return (orden.size - 1).coerceAtLeast(0)
    }

    /**
     * Los mensajes de una página. El aviso "Paciente activo: …" lo repite el bot en cada
     * activación; se deja solo el último para que no se acumule.
     */
    fun visibles(mensajes: List<MensajeHilo>, indice: Int): List<MensajeHilo> {
        val lista = if (orden.size <= 1) mensajes else mensajes.filter { it.episodio == orden[indice] }
        val ultimo = lista.indexOfLast(::esAvisoDeActivacion)
        if (ultimo < 0) return lista
        return lista.filterIndexed { i, mensaje -> i == ultimo || !esAvisoDeActivacion(mensaje) }
    }

    /** Solo en la consulta activa se escribe; las anteriores son de lectura. */
    fun esActiva(indice: Int, activo: String?): Boolean {
        if (orden.size <= 1) return true
        return orden[indice] == (activo ?: orden.last())
    }

    fun etiqueta(indice: Int, mensajes: List<MensajeHilo>): String {
        // Con una sola consulta, decirlo explica por qué las flechas no llevan a ningún lado.
        if (orden.size == 1) return "Única consulta"
        if (orden.isEmpty()) return "Sin consultas todavía"
        val fecha = mensajes.firstOrNull { it.episodio == orden[indice] }?.fecha
        val cuando = fecha?.let { " · ${Fechas.textoCorto(it)}" }.orEmpty()
        return "Consulta ${indice + 1}/${orden.size}$cuando"
    }

    companion object {
        /**
         * El autor sobre un mensaje ajeno, una vez por racha del mismo autor (como un grupo de
         * WhatsApp). `null` = no se muestra.
         */
        fun autor(lista: List<MensajeHilo>, indice: Int): String? {
            val mensaje = lista[indice]
            if (mensaje.propio) return null
            if (indice > 0) {
                val previo = lista[indice - 1]
                if (!previo.propio && previo.autorId == mensaje.autorId) return null
            }
            return if (mensaje.esBot) "Asistente" else mensaje.autorNombre ?: "Profesional"
        }

        fun esAvisoDeActivacion(mensaje: MensajeHilo): Boolean =
            mensaje.esBot && mensaje.contenido.trimStart().startsWith("Paciente activo:")
    }
}
